File: LeetCode/cpp/46.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int> &nums) {
        vector<int> visited(nums.size(), 0);
        vector<vector<int>> ret;
        vector<int> temp;
        BackTracking(nums, visited, temp, nums.size(), ret);
        return ret;
    }

    void BackTracking(vector<int> &nums, vector<int> &visited, vector<int> &temp, int n, vector<vector<int>> &ret) {
        if (n == 0) {
            ret.push_back(temp);
            return;
        }
        for (int i = 0; i < nums.size(); ++i) {
            if (visited[i] == 0) {
                visited[i] = 1;
                temp.push_back(nums[i]);
                BackTracking(nums, visited, temp, n - 1, ret);
                temp.pop_back();
                visited[i] = 0;
            }
        }
    }
};

File: LeetCode/python3/46.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        def BackTrack(m, per: list):
            if m == n:
                permutation.append(per)
                return per
            for i in range(n):
                if not visited[i]:
                    per.append(nums[i])
                    visited[i] = True
                    per = BackTrack(m + 1, per)
                    per = per[:-1]
                    visited[i] = False
            return per

        n = len(nums)
        visited = [False for _ in range(n)]
        per = []
        permutation = []
        BackTrack(0, [])
        return permutation
